src/WebJobs.Script.WebHost/WebJobsScriptHostService.cs

```diff
@@ -74,29 +78,32 @@ public async Task StartAsync(CancellationToken cancellationToken)
             }
         }
 
-        private async Task StartHostAsync(CancellationToken cancellationToken, int attemptCount = 0, bool skipHostJsonConfiguration = false)
+        private async Task StartHostAsync(CancellationToken cancellationToken, int attemptCount = 0, JobHostStartupMode startupMode = JobHostStartupMode.Normal)
         {
             cancellationToken.ThrowIfCancellationRequested();
 
             try
             {
                 bool isOffline = CheckAppOffline();
+                bool hasNonTransientErrors = startupMode.HasFlag(JobHostStartupMode.HandlingNonTransientError);
 
-                _host = BuildHost(isOffline, skipHostJsonConfiguration);
+                // If we're in a non-transient error state or offline, skip host initialization
+                bool skipJobHostStartup = isOffline || hasNonTransientErrors;
+
+                _host = BuildHost(skipJobHostStartup, skipHostJsonConfiguration: startupMode == JobHostStartupMode.HandlingConfigurationParsingError);
 
                 LogInitialization(_host, isOffline, attemptCount, _hostStartCount++);
 
                 await _host.StartAsync(cancellationToken);
 
-                // This means we had an error on a previous load, so we want to keep the LastError around
-                if (!skipHostJsonConfiguration)
+                if (!startupMode.HasFlag(JobHostStartupMode.HandlingError))
                 {
                     LastError = null;
-                }
 
-                if (!isOffline && !skipHostJsonConfiguration)
-                {
-                    State = ScriptHostState.Running;
+                    if (!isOffline)
+                    {
+                        State = ScriptHostState.Running;
```

Comment:
If we've skipped host startup due to an error, do we still want it to say Running?

Reply:
This is only hit if not in an error state. See the `if` above. I added a test that validates that `State` will be error and the exception will be set.